core/store/src/trie/mod.rs

```diff
@@ -416,6 +416,60 @@ pub struct Trie {
     pub counter: TouchedNodesCounter,
 }
 
+pub struct ShardTries {
+    tries: Vec<Arc<Trie>>,
+}
+
+impl ShardTries {
+    pub fn new(storage: Arc<Store>, num_shards: ShardId) -> Self {
+        let tries = (0..num_shards)
+            .map(|shard_id| Arc::new(Trie::new(Arc::clone(&storage), shard_id)))
+            .collect::<Vec<_>>();
+        ShardTries { tries }
+    }
+
+    pub fn new_trie_update(&self, shard_id: ShardId, state_root: CryptoHash) -> TrieUpdate {
+        TrieUpdate::new(self.get_trie_for_shard(shard_id), state_root)
+    }
+
+    pub fn get_trie_for_shard(&self, shard_id: ShardId) -> Arc<Trie> {
+        return Arc::clone(&self.tries[shard_id as usize]);
+    }
+
+    pub fn get_store(&self) -> Arc<Store> {
+        Arc::clone(
+            &self.tries[0].storage.as_caching_storage().expect("Always caching storage").store,
+        )
+    }
+
+    pub fn update_cache(&self, transaction: &DBTransaction) -> std::io::Result<()> {
+        let mut shards = vec![Vec::new(); self.tries.len()];
+        for op in &transaction.ops {
+            match op {
+                DBOp::Insert { col, ref key, ref value } if *col == ColState => {
+                    let shard_id = u64::from_le_bytes(key[0..8].try_into().unwrap());
+                    let hash = CryptoHash::try_from(&key[8..]).map_err(|_| {
+                        std::io::Error::new(ErrorKind::Other, "Key is always a hash")
+                    })?;
+                    shards[shard_id as usize].push((hash, Some(value.clone())));
+                }
+                DBOp::Delete { col, ref key } if *col == ColState => {
+                    let shard_id = u64::from_le_bytes(key[0..8].try_into().unwrap());
+                    let hash = CryptoHash::try_from(&key[8..]).map_err(|_| {
+                        std::io::Error::new(ErrorKind::Other, "Key is always a hash")
+                    })?;
```

Comment:
the code for getting shard_id and hash is complicated and should not be duplicated
please add a method `get_shard_hash_from_bytes`

Reply:
done